### PR TITLE
Protection for extrapolation of laser correction

### PR DESCRIPTION
**Pull request by fcouderc**
#### PR description:
Add a protection to avoir extrapolation of laser correction if time > t3 + delta t.
If the time of the event is higher than t3 (time of the "end" of the IOV) plus a safety margin then we force t = t3 + delta t.  Delta t is initialised to zero. This behaviour was enforced by adding spurious IOVs in the database. 

#### PR validation:
Check that delta t was properly propagated from the plugin EcalLaserCorrectionService to EcalLaserServiceDb the EcalLaserService and t_laser is properly set for events with t.< t3 and t > t3 + delta

### CODE REVIEW

**Comment by fcouderc** [comment]
@grasph @ferriff , please comment if necessary
PR for laser protection:   if ( t > t3 + delta t ) : t = t3 + delta t 
==> implying that for deltat = 0 (default value) then t > t3 ==> p = p3

**Comment by fcouderc** [comment]
FYI  @blenzi

File: CalibCalorimetry/EcalLaserCorrection/python/ecalLaserCorrectionService_cfi.py [hunk 1]
```diff
@@ -1,5 +1,9 @@
 import FWCore.ParameterSet.Config as cms
 
-EcalLaserCorrectionService = cms.ESProducer("EcalLaserCorrectionService")
+EcalLaserCorrectionService = cms.ESProducer("EcalLaserCorrectionService", 
+                                        ### delta t is the safety margin before stopping
+                                        ### the laser correction extrapolation (this is in seconds)
+                                        deltat_safety = cms.untracked.int32( 0  ) )
```
Comment (ferriff) [inline CalibCalorimetry/EcalLaserCorrection/python/ecalLaserCorrectionService_cfi.py:6]: probably this should be tracked, as it modifies the correction behaviour?

File: CalibCalorimetry/EcalLaserCorrection/plugins/EcalLaserCorrectionService.cc [hunk 2]
```diff
@@ -62,6 +65,8 @@ EcalLaserCorrectionService::EcalLaserCorrectionService(const edm::ParameterSet&
       .setConsumes(apdpnToken_)
       .setConsumes(linearToken_);
 
+  deltat_safety_ = fConfig.getUntrackedParameter<int>("deltat_safety", 0);
```
Comment (ferriff) [inline CalibCalorimetry/EcalLaserCorrection/plugins/EcalLaserCorrectionService.cc:68]: should be changed for tracked (?)
convert seconds to 64bit CMSSW format here:

```suggestion
  deltat_safety_ = fConfig.getParameter<int>("deltat_safety", 0);
  deltat_safety = deltat_safety<<32;
```

File: CalibCalorimetry/EcalLaserCorrection/src/EcalLaserDbService.cc [hunk 3]
```diff
@@ -12,8 +12,14 @@
 
 using namespace std;
 
+long long oneSecond = 4294967296;  // which is 2^32
```
Comment (ferriff) [inline CalibCalorimetry/EcalLaserCorrection/src/EcalLaserDbService.cc:15]: can be removed (see previous comment)

File: CalibCalorimetry/EcalLaserCorrection/src/EcalLaserDbService.cc [hunk 4]
```diff
@@ -192,6 +199,9 @@ float EcalLaserDbService::getLaserCorrection(DetId const& xid, edm::Timestamp co
     p_f = apdpnpair.p3;
   }
 
+  if (t > timestamp.t3.value() + (deltaT_safety_ * oneSecond))
+    t_laser = ((long long)timestamp.t3.value()) + deltaT_safety_ * oneSecond;
```
Comment (ferriff) [inline CalibCalorimetry/EcalLaserCorrection/src/EcalLaserDbService.cc:203]: the `* oneSecond` should be removed, either by converting seconds to `long long` 64bits CMSSW times with a 32 bit shifts as soon as the parameter is retrieved, or by doing a `<<32` shift here

File: CalibCalorimetry/EcalLaserCorrection/plugins/EcalLaserCorrectionServiceMC.cc [hunk 5]
```diff
@@ -56,6 +59,8 @@ EcalLaserCorrectionServiceMC::EcalLaserCorrectionServiceMC(const edm::ParameterS
       .setConsumes(apdpnToken_)
       .setConsumes(linearToken_);
 
+  deltat_safety_ = fConfig.getUntrackedParameter<int>("deltat_safety", 0);
```
Comment (ferriff) [inline CalibCalorimetry/EcalLaserCorrection/plugins/EcalLaserCorrectionServiceMC.cc:62]: same comment as for `EcalLaserCorrectionService.cc`

File: CalibCalorimetry/EcalLaserCorrection/src/EcalLaserDbService.cc [hunk 6]
```diff
@@ -164,6 +170,7 @@ float EcalLaserDbService::getLaserCorrection(DetId const& xid, edm::Timestamp co
   // interpolation
 
   edm::TimeValue_t t = iTime.value();
+  long long t_laser = t;
```
Comment (grasph) [inline CalibCalorimetry/EcalLaserCorrection/src/EcalLaserDbService.cc:173]: For code clarity I suggest to move this line to 201, so the definition of t_laser is grouped in one place.

**Review by grasph — COMMENTED** [review]
I suggest to call the new parameter "maxExtrapolationTime" or better "maxExtrapolationTimeInSec", that is more indicative than "deltat_safety". I agree with @ferriff that it should be a tracked parameter.

**Comment by fcouderc** [comment]
All comments taken into account...

**Comment by grasph** [comment]
+1

**Comment by pohsun** [comment]
please test

**Comment by fcouderc** [comment]
> -1
> 
> Tested at: [9e34140](https://github.com/cms-sw/cmssw/commit/9e34140476cf636f256a024fdf9820dc3e5d5baf)
> 
> CMSSW: CMSSW_11_2_X_2020-06-16-2300
> SCRAM_ARCH: slc7_amd64_gcc820
> You can see the results of the tests here:
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d9c36d/7122/summary.html
> 
> I found follow errors while testing this PR
> 
> **Failed tests: AddOn**
> 
> * **AddOn**:
> 
> I found errors in the following addon tests:
> 
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_2018.py realData=False globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:09:20 2020-date Wed Jun 17 10:00:01 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:05:52 2020-date Wed Jun 17 10:00:02 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:08:40 2020-date Wed Jun 17 10:00:04 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_2018.py realData=True globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:07:54 2020-date Wed Jun 17 10:00:11 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=False globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:06:47 2020-date Wed Jun 17 10:00:12 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=True globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:06:09 2020-date Wed Jun 17 10:00:18 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:07:25 2020-date Wed Jun 17 10:00:22 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=True globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:06:10 2020-date Wed Jun 17 10:00:26 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:09:36 2020-date Wed Jun 17 10:04:24 2020 s - exit: 18688
> cmsRun /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-16-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=False globalTag=@ inputFiles=@ : FAILED - time: date Wed Jun 17 10:09:47 2020-date Wed Jun 17 10:05:21 2020 s - exit: 18688

What shout I do about this ? The new tracked parameter is missing from HLTrigger/test/ pythons should I add there and commit ?

**Comment by grasph** [comment]
@fcouderc I think you need to add the new parameter in CalibCalorimetry/EcalLaserCorrection/python/ecalLaserCorrectionService_cfi.py . The file is loaded from Configuration/StandardSequences/python/AdditionalConditions_cff.py. Then, I suggest to run the cmsDriver command quoted in the HLT test python to test that the change is taken into account for the standard sequences.

I let @pohsun  answers for the For the HLT test python files.

**Comment by fcouderc** [comment]
@grasph not sure what you mean, the parameter is already in the service python otherwise nothing would work, do you have something else in mind ? 
@pohsun the problem is that the HLTrigger/Configuration/test fragments are implementing a new instance of the service rather than loading it from standard sequences. I suppose this behaviour is intended but I don't know if I should add the parameter there as well, or if they have an automated script to create these tests config. Please advise, this is really fast to do, I don't mind doing it.

**Comment by mmusich** [comment]
Hi @fcouderc, sorry to jump in, just to let you out of the *impasse*, direct modifications to the `HLTrigger/Configuration/test` fragments are forbidden by TSG policy. You should instead deploy a modifier for this file `HLTrigger/Configuration/python/customizeHLTforCMSSW.py` providing the new missing parameters in the menus which lack it. You can have a look to https://github.com/cms-sw/cmssw/pull/29959 or https://github.com/cms-sw/cmssw/pull/30046 or https://github.com/cms-sw/cmssw/pull/30060 to get an idea. By the way, to test your changes and automatically run all the fragments in `HLTrigger/Configuration/test` you can run `addOnTests.py`.
Hope this helps.

**Comment by grasph** [comment]
> @grasph not sure what you mean, the parameter is already in the service python otherwise nothing would work, do you have something else in mind ?

My bad. I had looked at the old version of the file.

**Comment by fcouderc** [comment]
@mmusich thanks a lot, I was assuming something like this indeed. This helps a lot. 
Can I do the PR myself for the customizeHLTforCMSSW.py or do I need to wait for somebody from TSG ?

**Comment by mmusich** [comment]
@fcouderc you can add a commit here. The PR needs to pass the tests to be accepted.

**Comment by fcouderc** [comment]
@mmusich I assume the intended behaviour in customizeHLTforCMSSW.py is that I only customise for my PR and comment out the previous customisation ? Is it indeed the case ?

**Comment by mmusich** [comment]
@fcouderc, no the customization is intended to be (semi-)permanent until TSG re-dumps the menu with your new patameters. You should leave all the other customizations as they are and only add your new customization function and apply it where relevant.

**Comment by fcouderc** [comment]
ok thanks @mmusich ! I realised that as well from the tests :) sorry for the spam

**Comment by christopheralanwest** [comment]
please test

**Comment by fcouderc** [comment]
I checked the failures which are due differences in the log. They all relates to wf 10224.0... I ran the workflows for both (with and w/o the PR). The differences are due to the opening of the PU file (with is different in the new PR so then it switches to the same one as in ref). I quote here the logs for the ref and new.
To me this does not seem related at all to the PR per se.

New (with PR):
```
24-Jun-2020 11:20:40 CEST  Initiating request to open file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/C134072E-F277-074E-9227-35C47E547F0C.root
24-Jun-2020 11:20:42 CEST  Successfully opened file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/C134072E-F277-074E-9227-35C47E547F0C.root
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 24-Jun-2020 11:20:55.577 CEST
24-Jun-2020 11:22:19 CEST  Closed file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/C134072E-F277-074E-9227-35C47E547F0C.root
24-Jun-2020 11:22:19 CEST  Initiating request to open file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/7D7EF45D-D1C1-8342-8C05-EDBE575C3364.root
24-Jun-2020 11:22:21 CEST  Successfully opened file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/7D7EF45D-D1C1-8342-8C05-EDBE575C3364.root
Begin processing the 2nd record. Run 1, Event 2, LumiSection 1 on stream 0 at 24-Jun-2020 11:22:54.244 CEST
Begin processing the 3rd record. Run 1, Event 3, LumiSection 1 on stream 0 at 24-Jun-2020 11:23:17.561 CEST
Begin processing the 4th record. Run 1, Event 4, LumiSection 1 on stream 0 at 24-Jun-2020 11:23:42.007 CEST
Begin processing the 5th record. Run 1, Event 5, LumiSection 1 on stream 0 at 24-Jun-2020 11:24:04.801 CEST
Begin processing the 6th record. Run 1, Event 6, LumiSection 1 on stream 0 at 24-Jun-2020 11:24:27.808 CEST
Begin processing the 7th record. Run 1, Event 7, LumiSection 1 on stream 0 at 24-Jun-2020 11:24:50.486 CEST
Begin processing the 8th record. Run 1, Event 8, LumiSection 1 on stream 0 at 24-Jun-2020 11:25:15.166 CEST
Begin processing the 9th record. Run 1, Event 9, LumiSection 1 on stream 0 at 24-Jun-2020 11:25:37.167 CEST
Begin processing the 10th record. Run 1, Event 10, LumiSection 1 on stream 0 at 24-Jun-2020 11:25:59.321 CEST
24-Jun-2020 11:26:21 CEST  Closed file file:step1.root
24-Jun-2020 11:26:27 CEST  Closed file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/7D7EF45D-D1C1-8342-8C05-EDBE575C3364.root

=============================================

MessageLogger Summary

 type     category        sev    module        subroutine        count    total
 ---- -------------------- -- ---------------- ----------------  -----    -----
    1 fileAction           -s MixingModule:mix                       3        3
    2 fileAction           -s MixingModule:mix                       2        2
    3 fileAction           -s MixingModule:mix                       1        1
    4 fileAction           -s file_close                             1        1
    5 fileAction           -s file_open                              2        2
```

ref (without PR):
```
24-Jun-2020 10:07:45 CEST  Initiating request to open file file:step1.root
24-Jun-2020 10:07:46 CEST  Successfully opened file file:step1.root
24-Jun-2020 10:07:47 CEST  Initiating request to open file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/7D7EF45D-D1C1-8342-8C05-EDBE575C3364.root
24-Jun-2020 10:07:50 CEST  Successfully opened file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/7D7EF45D-D1C1-8342-8C05-EDBE575C3364.root
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 24-Jun-2020 10:08:03.145 CEST
Begin processing the 2nd record. Run 1, Event 2, LumiSection 1 on stream 0 at 24-Jun-2020 10:09:38.611 CEST
Begin processing the 3rd record. Run 1, Event 3, LumiSection 1 on stream 0 at 24-Jun-2020 10:10:00.918 CEST
Begin processing the 4th record. Run 1, Event 4, LumiSection 1 on stream 0 at 24-Jun-2020 10:10:24.546 CEST
Begin processing the 5th record. Run 1, Event 5, LumiSection 1 on stream 0 at 24-Jun-2020 10:10:46.315 CEST
Begin processing the 6th record. Run 1, Event 6, LumiSection 1 on stream 0 at 24-Jun-2020 10:11:07.147 CEST
Begin processing the 7th record. Run 1, Event 7, LumiSection 1 on stream 0 at 24-Jun-2020 10:11:28.227 CEST
Begin processing the 8th record. Run 1, Event 8, LumiSection 1 on stream 0 at 24-Jun-2020 10:11:50.416 CEST
Begin processing the 9th record. Run 1, Event 9, LumiSection 1 on stream 0 at 24-Jun-2020 10:12:10.726 CEST
Begin processing the 10th record. Run 1, Event 10, LumiSection 1 on stream 0 at 24-Jun-2020 10:12:31.612 CEST
24-Jun-2020 10:12:52 CEST  Closed file file:step1.root
24-Jun-2020 10:12:54 CEST  Closed file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_10_6_0/RelValMinBias_13/GEN-SIM/106X_mc2017_realistic_v3-v1/10000/7D7EF45D-D1C1-8342-8C05-EDBE575C3364.root

=============================================

MessageLogger Summary

 type     category        sev    module        subroutine        count    total
 ---- -------------------- -- ---------------- ----------------  -----    -----
    1 fileAction           -s MixingModule:mix                       2        2
    2 fileAction           -s MixingModule:mix                       1        1
    3 fileAction           -s file_close                             1        1
    4 fileAction           -s file_open                              2        2
```

**Comment by fcouderc** [comment]
Difference in wf_1001
https://tinyurl.com/ybb8o2kk
https://tinyurl.com/yc5v34bu

which corresponds to the timestamp at which the WF was executed (i.e. 23/06). I believe this is expected to be different...

**Comment by fcouderc** [comment]
Summary of the 4 failures:
- reco comparison: 2 failures from wf_10224 related to edmLog, see https://github.com/cms-sw/cmssw/pull/30254#issuecomment-648723925
- DQM HistoTest: 4 failures:  
         - 2 in wf_10224.0: due to log https://github.com/cms-sw/cmssw/pull/30254#issuecomment-648723925
         - 1 in wf_1001.0: due to processTimeStamp (as I would expect):  https://github.com/cms-sw/cmssw/pull/30254#issuecomment-648742962
         - 1 in wf_8.0: not seen in DQM GUI comparisons (not sure why) but it seems to be related to the pixel DQM (only added a protection to Ecal laser correction) from https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-06-23-1100+d9c36d/37202/8.0_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS/

**Comment by Martin-Grunewald** [comment]
+1

**Comment by Martin-Grunewald** [comment]
please test

**Comment by fcouderc** [comment]
only failure remaining is in the pixel DQM... checking other PR it seems to be there as well.

**Comment by Martin-Grunewald** [comment]
It looks like the plugin should get a fillDescriptions method which will take care of the added parameters, with value 0.

**Comment by fcouderc** [comment]
@Martin-Grunewald , do you have an example of such a method ? Not sure what I should include at that point. Thx !

**Comment by fcouderc** [comment]
@Martin-Grunewald , I am a bit confused. Is the idea to have a default value to be = 0  ? Because the point of the untracked parameter was precisely the opposite ? Otherwise I can switch from track to untracked no (with default 0) ?

**Comment by Martin-Grunewald** [comment]
Well, 0 comes from this line in the cfi file:
```
maxExtrapolationTimeInSec = cms.uint32( 0 )
```
In a fillDescriprions method, you define all the parameter and also give a ('default') value to be used if that parameter is not in the py config of the module instance. This is an elegant way of adding new parameters.
There are lots of cases in CMMSW, eg https://github.com/cms-sw/cmssw/blob/master/Geometry/RPCGeometryBuilder/plugins/RPCGeometryESModule.cc#L66
This would then allow to remove the HLT customisation, which is the preferred way if a fillDescriptions can be easily added.

**Comment by fcouderc** [comment]
thx @Martin-Grunewald I will make another PR, I am not sure how to remove the HLT PR since it additionally got merged with another PR and I see that there will others adding on top. I will start from a fresh release and hopefully for the last time...

**Pull request closed without merging**